### DISCUSSION TITLE
Easy Pagination for Entity Framework in 3 steps

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#39] [General]
# Easy Pagination for Entity Framework in 3 steps

Pagination is the process of dividing a set into discrete pages. In the context of Entity Framework, that means we are only getting a certain amount of entries from the database.
And we will implement a very easy solution to make that happen in 3 steps. The result will look like this:
var pagedList = DbContext.BlogPosts.ToPagedList(page: 1, pageSize: 5);

https://steven-giesel.com/blogPost/09285b33-79e6-4879-95e0-35aeae5fbcc6

**Comment by VictorioBerra** [comment]
Thanks for sharing this. It's nice to see a detailed implementation with discussion around how/why choices were made.

Just curious, have your see this? https://github.com/dncuug/X.PagedList

**Reply by linkdotnet** [reply]
Hey, thanks for the tip.

Actually, yes, I do know that library. It was more or less the reason for that blog post. Initially, I used that library and you might see that the API is almost the same. I found it has some downsides:
 * It seems not really maintained anymore (in 1 year not much of commits)
 * It uses `Task.Run(() => ToList)` to mimic async operations. And I found there is a better way if you tailor-make the extension method to EF, RavenDb, ...
 * It seems a bit too complex for my use case.

So I let myself get inspired and did it on my own. For the blog itself, the use cases are manageable so I went down the road and implemented a very simple version on my own. Maybe someone can learn something from that, so I published the result :D

**Comment by piskov** [comment]
Every article using skip/take for pagination should contain big red disclaimer:

T H I S   W I L L   G E T   S L O W   A N D   Y O U ‘ L L   G E T   H U R T :-)

Why? Such code will work fast only while total number of suitable items in DB is not very big. Which is defcon-1 level threat, as problems will arise only when your app becomes used on daily basis for a while.

Suppose user will eventually get to a point where they have, for example, 1000 items and you need to show last 20.

`Skip(980).Take(20)` seems innocent enough until you realize that in order to get that 20, DB still must scan every single one of 980 before it.

So one should `skip` offset pagination and `take` (just couldn’t resist the pun) keyset pagination — with multiple pagination keys if needed — as soon as possible.

Some quickstarter links from .NET Data team:
- [Intro into keyset pagination](https://learn.microsoft.com/en-us/ef/core/querying/pagination#keyset-pagination)
- [Youtube discussion](https://youtu.be/DIKH-q-gJNU)
